Server/src/services/vehicle.service.ts

```diff
@@ -358,138 +466,148 @@ export default class VehicleService {
-		}
-	}
-
-	/**
-	 * This is just a wrapper that gets the speed of the tracker assigned to a given vehicle. Also it accumulates all
-	 * tracker data as a vehicle could have more than one tracker assigned.
-	 * @param vehicle `Vehicle` to get the speed for
-	 * @returns last known speed (always a positive number) of `vehicle` based on tracker data, -1 if speed is unknown
-	 */
-	public static async getVehicleSpeed(vehicle: Vehicle): Promise<number> {
-		// TODO: instead of just returning last heading a computation / validation could be better (e.g. for multiple trackers)
-
-		// get last heading of logs
-		const logs = await database.logs.getAll(vehicle.uid)
-		if (logs.length == 0) {
-			return -1
-		}
-		return logs[0].speed
-	}
+    /**
+     * Compute average heading of all trackers assigned to a specified vehicle. 
+     * No headings from app will be used here due to mobility.
+     * @param vehicle `Vehicle` to get the heading for
+     * @returns average heading (between 0 and 359) of `vehicle` based on tracker data, -1 if heading is unknown
+     */
+    public static async getVehicleHeading(vehicle: Vehicle): Promise<number>{
+        // get all trackers for given vehicle
+        const trackers = await database.trackers.getByVehicleId(vehicle.uid)
+        if (trackers.length == 0) {
+            logger.error(`No tracker found for vehicle ${vehicle.uid}.`)
+            return -1
+        }
+
+        // get all last known tracker logs
+        let lastLogs: Log[] = []
+        for (let i = 0; i < trackers.length; i++) {
+
+            // get last log entry for this tracker
+            const lastLog = await database.logs.getAll(vehicle.uid, trackers[i].uid, 1)
+            if (lastLog.length != 1) {
+                // just try other trackers if there are no logs for this tracker
+                logger.warn(`Did not find any entry for vehicle ${vehicle.uid} and tracker ${trackers[i].uid}.`)
+                continue
+            }
+            lastLogs.push(lastLog[0])
+        }
+
+        // check if we got any log
+        if (lastLogs.length == 0) {
+            logger.error(`Could not find any tracker log for vehicle ${vehicle.uid}`)
+            return -1
+        }
+
+        // actually computing average
+        let avgHeading = 0
+        for (let i = 0; i < lastLogs.length; i++) {
+            avgHeading += lastLogs[i].heading / lastLogs.length
+        }
+        return avgHeading
+    }
+
+    /**
+     * Determine heading of a vehicle related to its track (either "forward" or "backward")
+     * @param vehicle `Vehicle` to get the heading for
+     * @param trackKm track kilometer at which the vehicle currently is (can be found with `VehicleService.getVehicleTrackDistanceKm`)
+     * @returns 1 or -1 if the vehicle is heading towards the end and start of the track respectively, 0 if heading is unknown
+     */
+    public static async getVehicleTrackHeading(vehicle: Vehicle, trackKm: number): Promise<number>{
```

Comment:
This is the difference to PR #50. Because I expect the track kilometer here already, I do not need to call the related function, which led to the infinite recursion. The track kilometer can be accessed easily with `getVehicleTrackDistanceKm`. All calling sites have been adjusted.